src/main/java/org/embulk/input/marketo/rest/MarketoRestClient.java

```diff
@@ -0,0 +1,335 @@
+        String getClientId();
+
+        @Config("marketo_limit_interval_milis")
+        @ConfigDefault("20000")
+        Integer getMarketoLimitIntervalMilis();
+
+        @Config("batch_size")
+        @ConfigDefault("300")
+        Integer getBatchSize();
+
+
+    }
+
+    public MarketoRestClient(PluginTask task, Jetty92RetryHelper retryHelper)
+    {
+        this(task.getEndpoint(), task.getIdentityEndpoint(), task.getClientId(), task.getClientSecret(), task.getBatchSize(), task.getMarketoLimitIntervalMilis(), retryHelper);
+    }
+
+    public MarketoRestClient(String endPoint, String identityEndPoint, String clientId, String clientSecret, Integer batchSize, int marketoLimitIntervalMilis, Jetty92RetryHelper retryHelper)
+    {
+        super(identityEndPoint, clientId, clientSecret, marketoLimitIntervalMilis, retryHelper);
+        this.endPoint = endPoint;
+        this.batchSize = batchSize;
+    }
+
+    public List<Column> describeLead()
+    {
+        MarketoResponse<ObjectNode> jsonResponse = doGet(endPoint + MarketoRESTEndpoint.DESCRIBE_LEAD.getEndpoint(), null, null, new MarketoResponseJetty92EntityReader<ObjectNode>(READ_TIMEOUT_MILLIS));
+        List<Column> columnsList = new ArrayList<>();
+        List<ObjectNode> fields = jsonResponse.getResult();
+        for (int i = 0; i < fields.size(); i++) {
+            ObjectNode field = fields.get(i);
+            String dataType = field.get("dataType").asText();
+            ObjectNode restField = (ObjectNode) field.get("rest");
+            String name = restField.get("name").asText();
+            columnsList.add(new Column(i, name, getType(dataType)));
+        }
+        return columnsList;
+    }
+
+    private Type getType(String dataType)
+    {
+        return TYPE_MAPPING.containsKey(dataType.toLowerCase()) ? TYPE_MAPPING.get(dataType.toLowerCase()) : Types.STRING;
+    }
+
+    public String createLeadBulkExtract(Date startTime, Date endTime, List<String> extractFields)
+    {
+        SimpleDateFormat timeFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
+        MarketoBulkExtractRequest marketoBulkExtractRequest = new MarketoBulkExtractRequest();
+        marketoBulkExtractRequest.setFields(extractFields);
+        marketoBulkExtractRequest.setFormat("CSV");
+        Map<String, MarketoFilter> filterMap = new HashMap<>();
+        DateRangeFilter dateRangeFilter = new DateRangeFilter();
+        dateRangeFilter.setStartAt(timeFormat.format(startTime));
+        dateRangeFilter.setEndAt(timeFormat.format(endTime));
+        filterMap.put("createdAt", dateRangeFilter);
+        marketoBulkExtractRequest.setFilter(filterMap);
+        return sendCreateBulkExtractRequest(marketoBulkExtractRequest, MarketoRESTEndpoint.CREATE_LEAD_EXTRACT);
+    }
+
+    public String createActitvityExtract(Date startTime, Date endTime, List<String> activityTypes)
+    {
+        SimpleDateFormat timeFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
```

Comment:
Seems to be duplicated with the code above. The difference is only how the `filterMap` is built & the real api call.